crates/dbs-miniball/docs/CPUVirtualization.md

```diff
@@ -0,0 +1,869 @@
+# CPU 虚拟化
+
+## 概述
+在配置 KVM、guest memory 后，开始配置 vCPU。通过 dbs-boot、dbs-arch、kvm-ioctls crates 完成。
+
+1. dbs-arch：包含了 CPU 架构相关的仓库，例如 CPUID、MSR 寄存器的定义，CPU 虚拟化过程中会用到这些寄存器 / CPUID。
+1. dbs-boot：启动过程中会用到的仓库，其中 MPTable 会和初始 CPU 个数等信息相关。
+1. kvm-ioctls：其中会涉及到 create_vcpu 等 CPU 虚拟化相关 ioctl 以及 vCPU 相关 fd vcpuf。
+
+## 步骤
+创建 KvmVm 时，配置 MPTable，创建 KVM irqchip，创建 vCPU。
+
+1. 配置 MPTable。告诉 Guest 操作系统多处理器配置，即使使用单个 vCPU 也是必需的。
+1. 创建 KVM irqchip。创建虚拟 IOAPIC 和虚拟 PIC，并为 LAPIC 设置未来的 vCPU。
+1. 创建 vCPU。为每个 vCPU 向 KVM 注册一个 fd。
+   1. 配置 CPUID。必须。因为它是 Guest 识别自身是被虚拟化的方式。
+   1. 配置 MSR。这些寄存器控制（除其他外）处理器功能。
+   1. 配置其他寄存器（kvm_regs、kvm_sregs、fpu）和 LAPIC。
+
+## MPTable
+##### 原理介绍
+对称多处理器（Symmetrical Multi-Processing），简称 SMP，指在一个计算机上汇集了一组处理器。在这种架构中，一台计算机由多个处理器组成，所有的处理器都可以平等地访问内存、I/O 和外部中断，运行操作系统的单一副本， 并共享内存和其他资源。操作系统将任务均匀地分布在多个 CPU 中，从而极大地提高了整个系统的数据处理能力。在虚拟 SMP 系统时，每个 CPU 使用一个线程来模拟，如下图所示。
+
+![SMP.png](img/SMP.png)
+> From：深度探索Linux系统虚拟化原理与实现
+
+
+其中有两个主要部分需要考虑：一是 KVM 需要把这些 VCPU 的信息告知 Guest，这样 Guest 才可以充分利用多处理器资源；二是多处理器系统只能由一个处理器准备基础环境，这些环境准备工作如果由多个处理器不加保护地并发执行，将会带来灾难，此时其他处理器都必须处于停止状态，当基础环境准备好后，其他处理器再启动运行。
+
+为了获得处理器信息，操作系统有两种获取处理器信息的方式：一种是 Intel 的 Multi Processor Specification（后续简称 MP Spec）约定的方式；另外一种是 ACPI MADT（Multiple APIC Description Table）约定的方式。MP Spec 约定的核心数据结构包括两部分：MP Floating Pointer Structure（后续简称 MPF）和 MP  Configuration Table（后续简称 MP Table）的地址，如下图所示。
+
+![MPtable.png](img/MPTable.png)
```

Comment:
Redraw the picture by yourself and add not only the picture but the original file is better.

Reply:
OK, I'll revise it later.